File: printBinaryTree/printBinaryTree.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <cmath>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    vector<vector<string>> printTree(TreeNode* root) {
        int h = getHeight(root);
        int w = pow(2, h) - 1;
        vector<vector<string> > res(h, vector<string>(w, ""));
        queue<tuple<TreeNode*, int, int> > q;
        q.emplace(root, 0, w - 1);
        int level = 0;
        while (!q.empty()) {
            int q_size = q.size();
            for (int i = 0; i < q_size; i++) {
                tuple<TreeNode*, int, int> t = q.front();
                q.pop();
                TreeNode* cur = get<0>(t);
                int left_bound = get<1>(t), right_bound = get<2>(t);
                int pos = (left_bound + right_bound) / 2;
                res[level][pos] = to_string(cur->val);
                if (cur->left) {
                    q.emplace(cur->left, left_bound, pos - 1);
                }
                if (cur->right) {
                    q.emplace(cur->right, pos + 1, right_bound);
                }
            }
            level++;
        }
        return res;
    }
    int getHeight(TreeNode* root) {
        if (!root)
            return 0;
        return max(getHeight(root->left), getHeight(root->right)) + 1;
    }
};

int main()
{
    Solution s;
    int tree[7] = {3,9,20,9999,9999,15,7};
    int size = sizeof(tree)/sizeof(tree[0]);
    vector<TreeNode *> vec;
    for (int i = 0; i < size; i++) {
        if (tree[i] != 9999) {
            vec.push_back(new TreeNode(tree[i]));
        } else {
            vec.push_back(NULL);
        }
    }
    for (int i = 0; i < size/2; i++) {
        if(!vec[i])
            continue;
        if (i*2 + 1 < size)
            vec[i]->left = vec[i*2 + 1];
        if (i*2 + 2 < size)
            vec[i]->right = vec[i*2 + 2];
    }
    for (auto& v : s.printTree(vec[0])) {
        for (string& ss : v) {
            if (!ss.empty())
                cout << "\""<< ss << "\",";
            else
                cout << "\"\",";
        }
        cout << endl;
    }
    for (TreeNode *t : vec) {
        delete t;
    }
    vec.clear();
    return 0;
}

File: 655.printBinaryTree/printBinaryTree.java
import java.util.*;
class TreeNode {
	int val;
	TreeNode left;
	TreeNode right;
	TreeNode(int x) { val = x; }
}
public class PrintBinaryTree {
    class Params {
        Params(TreeNode _root, int _l, int _r) {
            root = _root;
            l = _l;
            r = _r;
        }
        TreeNode root;
        int l, r;
    }
    public List<List<String>> printTree(TreeNode root) {
        int h = getHeight(root);
        int w = (int) Math.pow(2, h) - 1;
        List<List<String>> res = new ArrayList<>();
        String[][] ans = new String[h][w];
        for (String[] arr : ans)
            Arrays.fill(arr, "");
        bfs(ans, root, w);
        for (String[] arr : ans)
            res.add(Arrays.asList(arr));
        return res;
            
    }
    public void bfs(String[][] ans, TreeNode root, int w) {
        Queue<Params> q = new LinkedList();
        q.add(new Params(root, 0, w - 1));
        int level = 0;
        while (!q.isEmpty()) {
            int qSize = q.size();
            for (int i = 0; i < qSize; i++) {
                Params p = q.remove();
                int pos = (p.l + p.r) / 2;
                ans[level][pos] = "" + p.root.val;
                if (p.root.left != null)
                    q.add(new Params(p.root.left, p.l, pos - 1));
                if (p.root.right != null)
                    q.add(new Params(p.root.right, pos + 1, p.r));
            }
            level++;
        }
    }
    public int getHeight(TreeNode root) {
        if (root == null) return 0;
        return Math.max(getHeight(root.left), getHeight(root.right)) + 1;
    }
	public static void main(String[] args) {
		int tree[] = {4,2,7,1,3,6,9};
		ArrayList<TreeNode> arrayList = new ArrayList<>();
		for (int i = 0; i < tree.length; i++) {
			if (tree[i] != 9999)
				arrayList.add(new TreeNode(tree[i]));
			else
				arrayList.add(null);
		}
		for (int i = 0; i < tree.length / 2; i++) {
			if (arrayList.get(i) == null)
				continue;
			if (i * 2 + 1 < tree.length)
				arrayList.get(i).left = arrayList.get(i * 2 + 1);
			if (i * 2 + 2 < tree.length)
				arrayList.get(i).right = arrayList.get(i * 2 + 2);
		}
		PrintBinaryTree s = new PrintBinaryTree();
		for (List<String> arr : s.printTree(arrayList.get(0))) {
			for (String ss : arr) {
				if (ss.equals("")) {
					System.out.print("\"\",");
				} else { 
					System.out.print("\"" + ss + "\",");
				}
			}
			System.out.println();
		}
	}
}
